### PR TITLE
Add default traits and setter

### PR DESCRIPTION
**Pull request by abayer**
I'm honestly not sure why these weren't already here.

cc @rsandell @dwnusbaum

### CODE REVIEW

**Comment by abayer** [comment]
@stephenc @rsandell - could one of you review and hopefully merge/release this? Thanks.

File: src/test/java/jenkins/scm/impl/mock/MockSCMSource.java [hunk 1]
```diff
@@ -362,6 +369,8 @@ public ListBoxModel doFillRepositoryItems(@QueryParameter String controllerId) t
             return SCMSourceTrait._for(this, MockSCMSourceContext.class, MockSCMBuilder.class);
         }
 
+        @Override
+        @Nonnull
```
Comment (stephenc) [inline src/test/java/jenkins/scm/impl/mock/MockSCMSource.java]: Wrong Nonnull annotation

File: src/main/java/jenkins/scm/api/SCMNavigator.java [hunk 2]
```diff
@@ -37,6 +37,10 @@
 import hudson.model.TaskListener;
 import hudson.util.AlternativeUiTextProvider;
 import hudson.util.LogTaskListener;
+import jenkins.model.TransientActionFactory;
+import jenkins.scm.api.trait.SCMTrait;
```
Comment (stephenc) [inline src/main/java/jenkins/scm/api/SCMNavigator.java]: A-Z no grouping

File: src/main/java/jenkins/scm/api/SCMNavigatorDescriptor.java [hunk 3]
```diff
@@ -27,13 +27,16 @@
 import edu.umd.cs.findbugs.annotations.CheckForNull;
 import edu.umd.cs.findbugs.annotations.NonNull;
 import hudson.model.Descriptor;
-import java.util.Collections;
-import java.util.LinkedHashSet;
-import java.util.Set;
```
Comment (stephenc) [inline src/main/java/jenkins/scm/api/SCMNavigatorDescriptor.java]: A-Z no grouping

File: src/test/java/jenkins/scm/impl/mock/MockSCMNavigator.java [hunk 4]
```diff
@@ -52,6 +46,13 @@
 import org.jenkinsci.Symbol;
 import org.kohsuke.stapler.DataBoundConstructor;
```
Comment (stephenc) [inline src/test/java/jenkins/scm/impl/mock/MockSCMNavigator.java]: A-Z no grouping

File: src/test/java/jenkins/scm/impl/mock/MockSCMNavigator.java [hunk 5]
```diff
@@ -52,6 +46,13 @@
...
 
+import javax.annotation.Nonnull;
```
Comment (stephenc) [inline src/test/java/jenkins/scm/impl/mock/MockSCMNavigator.java]: Where is the published JSR defining a spec for a `javax` annotation of this name?

**Comment by abayer** [comment]
Imports and annotation updated.

**Comment by abayer** [comment]
@stephenc ping for re-review?

**Comment by stephenc** [comment]
Did you address my query on symbols generically?

**Comment by abayer** [comment]
No, because that seems to be a separate thing from just adding these methods.

**Comment by stephenc** [comment]
I think it would be a mistake to significantly push traits further after discovering the @symbol name conflict issue without finding a way for two sources to have different implications of the same trait symbol name and not have the symbol support blow up and force use of the FQCN despite the case that the implications are unambiguously assignable

**Comment by stephenc** [comment]
If you are critically blocked on this, then go ahead... but my belief is the symbol problem is intrinsically linked (as once I hit it I stopped pushing beyond git based SCMs (which at least share a common parent class)

**Comment by abayer** [comment]
I'm definitely not disagreeing re: there being an issue with symbol resolution - I'm just saying that adding no-op traits getter/setter methods doesn't actually create any problems. I need this specifically to be able to do instantiation of a multibranch project with an arbitrary `SCMSource`, and since you can't actually create a functional multibranch project using `GitSCMSource` et al without manually setting the default traits, my alternative to this PR is to do reflection crap. So yeah, we do need to move forward on this - either it's gonna get merged/released ASAP or I need to change my code elsewhere to do reflection ugliness.

**Comment by stephenc** [comment]
Bitbucket and GitHub (And Gitea And in future GitLab) all want their own “discover fork PRs” trait with different options and backing implementation.

We’d like each one to be able to call the trait the same name, eg `discoverForkPRs` but as soon as we do that you cannot instantiate the trait using symbol... (try configuring the `resolveScm` step in the Pipeline Syntax Web UI)

So instead, we end up with `discoverGitHubForkPRs`, `discoverBitbucketForkPRs`, etc... which is not good.

I think a “symbol helper” extension of some sort that can filter the extension list contextually might be the solution... that way scm-api could register one and use the bag of trait descriptors to limit the options and we’d only have the GitHub option for GitHub so it could “share” the same trait name for the equivalent functionality

**Comment by stephenc** [comment]
Speaking of bag of trait descriptors (I’m on phone and cannot check) you do add the trait descriptors bag to the abstract SCMSourceDescriptor base class?

**Comment by abayer** [comment]
Yes, I did. Let me know when you've created the JIRA.

**Comment by stephenc** [comment]
You may be waiting some time if you’re waiting on me to create the jira

**Comment by abayer** [comment]
Fine by me - if you're waiting for me to do it, you're gonna wait even longer, since you're the one who actually knows what you want. =)

**Comment by stephenc** [comment]
sad panda
https://issues.jenkins-ci.org/browse/JENKINS-53825

**Comment by abayer** [comment]
@stephenc So are we ok to merge this, then?

**Comment by stephenc** [comment]
If @rsandell agrees

File: src/main/java/jenkins/scm/api/SCMNavigator.java [hunk 6]
```diff
@@ -158,6 +159,22 @@ protected final void resetId() {
...
+     * @param traits the list of traits
+     */
+    public void setTraits(@CheckForNull List<SCMTrait<? extends SCMTrait<?>>> traits) {
+    }
+
+    /**
+     * Gets the traits for this navigator.
+     * @return traits the list of traits, empty by default.
+     */
+    @CheckForNull
```
Comment (rsandell) [inline src/main/java/jenkins/scm/api/SCMNavigator.java]: Eeh, in the docs it says `@NonNull`?

Comment (abayer) [reply src/main/java/jenkins/scm/api/SCMNavigator.java]: whoops! Fixing.

**Pull request merged**